src/rust/engine/graph/src/node.rs

```diff
@@ -37,7 +37,7 @@ pub trait Node: Clone + Debug + Display + Eq + Hash + Send + 'static {
   ///
   /// If the node result is cacheable, return true.
   ///
-  fn cacheable(&self, context: &Self::Context) -> bool;
+  fn cacheable(&self) -> bool;
```

Comment:
This change is surprising, and looks like it required a lot of refactoring of the tests... it also removes some flexibility, because I could imagine needing to use the context to determine whether a Node should be cached (as the tests did).

If this was intended to make it easier to make the change to `Graph::walk`, it might be better to go the predicate route?

Reply:
>I could imagine needing to use the context to determine whether a Node should be cached (as the tests did).

It seems like the only reason this trait method was taking a context was to make the graph tests simpler by not needing to store cacheability information on the TNode. I can't think of a reason why we couldn't store some unknown future state needed to determine a nodes cacheability on the real NodeKey, like we do currently with the Task cacheability field.
I considered keeping the trait method signature the same, but I couldn't figure out a good way to have a Context available from either the scheduler, the watcher or the graph. The latter two of which exist as fields on the Context. That seems like it would require much larger refactoring of the source to resolve. I would rather prioritize simplicity in the code over simplicity of the test harness. The combination of those two things made me go this route. I'm fairly happy with how little I had to change the way TNodes are instantiated.

>If this was intended to make it easier to make the change to Graph::walk, it might be better to go the predicate route?

I see these as separate concerns. Even if we go the predicate route we still need a reference to the context in the watcher and scheduler, or Graph, which I can't see a way of doing without lots of refactors. The predicate issue is more about building flexibility into Graph::walk which I can see the benefit of.